libs/editor/egui_editor/src/events.rs

```diff
@@ -1,97 +1,306 @@
+                //                        range of replaced text:      |<--->|
+                //          range of subsequent cursor selection:  |<--->|
+                //                        text after replacement: * * * X *
+                // adjusted range of subsequent cursor selection:  |<->|
+                if let Some(selection_origin) = mod_cursor.selection_origin {
+                    if mod_cursor.pos < selection_origin {
+                        mod_cursor.selection_origin = Some(replaced_text_range.start);
                     } else {
-                        self.cursor.selection_origin = None;
+                        mod_cursor.pos = replaced_text_range.start;
                     }
-                    if modifiers.command {
-                        self.cursor.loc = self.last_cursor_position();
-                        let (galley_idx, _) = self.galley_and_cursor();
-                        self.fix_cursor(galley_idx, false);
-                        self.cursor.x_target = None;
-                        break;
+                } else {
+                    panic!("this code should be unreachable")
+                }
+            }
+            (false, true) => {
+                // case 5:
+                //                       text before replacement: * * * * * * *
+                //                        range of replaced text:  |<------->|
+                //          range of subsequent cursor selection:    |<--->|
+                //                        text after replacement: * X *
+                // adjusted range of subsequent cursor selection:    |
+                mod_cursor.pos = replaced_text_range.end;
+                mod_cursor.pos = mod_cursor.pos + text_replacement_len - replaced_text_len;
+                mod_cursor.selection_origin = Some(mod_cursor.pos);
+            }
+            (true, false) => {
+                // case 6:
+                //                       text before replacement: * * * * * * *
+                //                        range of replaced text:    |<--->|
+                //          range of subsequent cursor selection:  |<------->|
+                //                        text after replacement: * * X * *
+                // adjusted range of subsequent cursor selection:  |<--->|
+                if let Some(selection_origin) = mod_cursor.selection_origin {
+                    if mod_cursor.pos < selection_origin {
+                        mod_cursor.selection_origin =
+                            Some(selection_origin + text_replacement_len - replaced_text_len);
+                    } else {
+                        mod_cursor.pos = mod_cursor.pos + text_replacement_len - replaced_text_len;
                     }
+                } else {
+                    panic!("this code should be unreachable")
+                }
+            }
+        }
+    }
+}
+
+fn calc_modifications<'a>(
```

Comment:
existing event processing code, updated to produce modifications instead of apply them